src/dataflow/src/source/mod.rs

```diff
@@ -849,6 +855,221 @@ enum MessageProcessing {
+        self.sender
+            .send(Event::Message(*self.timestamp, msg))
+            .await
+            .or_else(|_| Err(anyhow!("channel closed")))
+    }
+
+    /// Record a deletion of a row in the current transaction
+    pub async fn delete(&self, row: Row) -> anyhow::Result<()> {
+        let msg = Ok((row, *self.timestamp, -1));
+        self.sender
+            .send(Event::Message(*self.timestamp, msg))
+            .await
+            .or_else(|_| Err(anyhow!("channel closed")))
+    }
+}
+
+/// A thread-safe transaction manager that is responsible for assigning timestamps to the
+/// transactions submitted to the system. Rows can be inserted individually using the
+/// [insert](Timestamper::insert) and [delete](Timestamper::delete) methods or as part of a bigger
+/// transaction.
+///
+/// When a transaction is started using [start_tx](Timestamper::start_tx) the internal clock will be
+/// frozen and any subsequent rows will be timestamped with the exact same timestamp. The
+/// transaction is committed automatically as soon as the transaction object gets dropped.
+pub struct Timestamper {
+    inner: Arc<RwLock<Timestamp>>,
+    sender: EventSender,
+    tick_duration: Duration,
+}
+
+impl Timestamper {
+    fn new(sender: EventSender, tick_duration: Duration) -> Self {
+        let now = UNIX_EPOCH
+            .elapsed()
+            .expect("system clock before 1970")
+            .as_millis()
+            .try_into()
+            .expect("materialize has existed for more than 500M years");
+        Self {
+            inner: Arc::new(RwLock::new(now)),
+            sender,
+            tick_duration,
+        }
+    }
+
+    /// Start a transaction at a particular point in time. The timestamper will freeze its internal
+    /// clock while a transaction is active.
+    pub async fn start_tx<'a>(&'a self) -> Transaction<'a> {
+        Transaction {
+            timestamp: self.inner.read().await,
+            sender: &self.sender,
+        }
+    }
+
+    /// Record an insertion of a row
+    pub async fn insert(&self, row: Row) -> anyhow::Result<()> {
+        self.start_tx().await.insert(row).await
+    }
+
+    /// Record a deletion of a row
+    pub async fn delete(&self, row: Row) -> anyhow::Result<()> {
+        self.start_tx().await.delete(row).await
+    }
+
+    /// Records an error. After this method is called the source will be permanently in an errored
+    /// state
+    pub async fn error(&self, err: SourceError) -> anyhow::Result<()> {
+        let timestamp = self.inner.read().await;
+        self.sender
+            .send(Event::Message(*timestamp, Err(err)))
+            .await
+            .or_else(|_| Err(anyhow!("channel closed")))
+    }
+
+    /// Attempts to monotonically increase the current timestamp and provides a Progress message to
+    /// timely.
+    ///
+    /// This method will wait for all current transactions to commit before advancing the clock and
```

Comment:
Thinking through deadlocks here. What happens if a simple source impl grabs a tx, a tick happens (and thus waits for the tx to drop), but then the simple source impl grabs another tx before dropping the first? I think it deadlocks? Should we care about detecting that, document it, or trying to prevent it in some way?

Reply:
Hm, yeah you're right, it will deadlock in this case. I had originally made `start_tx` have a `&mut self` receiver but that made using it harder. I will revisit the design though to see if we can somehow avoid the issue or detect it as you say. Glad you like the new interface!